### ISSUE TITLE
http/http responseNode are bBroken on Startup

### ISSUE THREAD

**Issue by Ryonez** [node-red/node-red#2325]
### What are the steps to reproduce?

Set up a http/http response set. The one I use retrieves data or triggers other nodes based on the query.

### What happens?
When Node-RED starts, either from a cold start or a restart, the http node parrots back the query, instead of responding properly.

### What do you expect to happen?
For Node-RED to process the query and respond accordingly.

### Is there a known workaround?
Making any change to the flow and deploying it seems to kick it into gear. For me, I've been toggling a debug node on and off to get it going for now. This still requires manual intervention whenever I start Node-RED however.

### Please tell us about your environment:

- [x] Node-RED version: 0.20.8
- [x] Node.js version: I don't know (Using docker image: nodered/node-red-docker:v8)
- [x] npm version: I don't know (Using docker image: nodered/node-red-docker:v8)
- [x] Platform/OS: Unraid 6.7.2
- [x] Browser: Firefox 69

**Comment by knolleary** [comment]
Can you share your actual flow? Without that, there's not much to go on.

**Comment by Ryonez** [comment]
@knolleary 
Here's the relevant nodes relating to http requests:
https://pastebin.com/VneWL60D

**Comment by knolleary** [comment]
Given how widely the HTTP In/Response nodes are used, we've not had any similar reports to this, nor can I reproduce it locally.

Can you add a Debug node after the HTTP In node to check the node is handling the request or get any further insight into what might be happening?

**Comment by Ryonez** [comment]
I stuck a debug node in, and restarted the container.

I'm getting the query parroted back again, and the debug node isn't reporting anything.

**Comment by knolleary** [comment]
After you restart the container, have you reloaded the editor to see what flows it thinks are deployed?

**Comment by Ryonez** [comment]
I did, and redid the test. No changes.

**Comment by knolleary** [comment]
How are you running the container? Have you mounted a volume to `/data` to store your flows?

**Comment by Ryonez** [comment]
Yes, it's all persistent. It just for some reason requires a redeploy on the flow to start this.

**Comment by knolleary** [comment]
Can you edit your settings file and increase the logging level to trace. That will cause NR to log a lot more information about the nodes/flows it is starting. You can then check that against the flow configuration you have deployed to see if it is starting the nodes you think it ought to be.

But to be clear, when you load the editor it shows the flows that are running. There is no way for it to show a different set of flows.

Given we can't recreate this, we're going to need you to do some digging to see if you can identify what is going on.

**Comment by Ryonez** [comment]
Sure, I'll make the changes and have a look.

**Comment by Ryonez** [comment]
```
14 Oct 23:47:23 - [trace] [flow:6f4ddae8.18649c] ------------------|--------------|-----------------
14 Oct 23:47:23 - [trace] [flow:1f453d6c.db47e3] start flow
14 Oct 23:47:23 - [trace] [flow:1f453d6c.db47e3] ------------------|--------------|-----------------
14 Oct 23:47:23 - [trace] [flow:1f453d6c.db47e3] id | type | alias
14 Oct 23:47:23 - [trace] [flow:1f453d6c.db47e3] ------------------|--------------|-----------------
14 Oct 23:47:23 - [trace] [flow:1f453d6c.db47e3] 284ebc08.f52484 | http in |
14 Oct 23:47:23 - [trace] [flow:1f453d6c.db47e3] a76a4504.9db768 | http response |
14 Oct 23:47:23 - [trace] [flow:1f453d6c.db47e3] a97f6010.fc38 | mongodb out |
14 Oct 23:47:23 - [trace] [flow:1f453d6c.db47e3] a699685b.9d3fe | inject |
14 Oct 23:47:23 - [trace] [flow:1f453d6c.db47e3] bbfbff00.726e4 | function |
14 Oct 23:47:23 - [trace] [flow:1f453d6c.db47e3] 4e3a663c.743e4 | inject |
14 Oct 23:47:23 - [trace] [flow:1f453d6c.db47e3] ed541487.d9036 | mongodb in |
14 Oct 23:47:23 - [trace] [flow:1f453d6c.db47e3] d95674c7.2e16c8 | debug |
14 Oct 23:47:23 - [trace] [flow:1f453d6c.db47e3] 51e47094.51547 | function |
14 Oct 23:47:23 - [trace] [flow:1f453d6c.db47e3] 43c32cdf.637a74 | inject |
14 Oct 23:47:23 - [trace] [flow:1f453d6c.db47e3] 547b02e0.2b4abc | mongodb out |
14 Oct 23:47:23 - [trace] [flow:1f453d6c.db47e3] 6ae9ee19.621b1 | json |
14 Oct 23:47:23 - [trace] [flow:1f453d6c.db47e3] 2d41bf64.e519e | switch |
14 Oct 23:47:23 - [trace] [flow:1f453d6c.db47e3] fd6ae544.7c3ae8 | switch |
14 Oct 23:47:23 - [trace] [flow:1f453d6c.db47e3] 6171af98.b8151 | link in |
14 Oct 23:47:23 - [trace] [flow:1f453d6c.db47e3] a9d54cc5.89046 | mongodb out |
14 Oct 23:47:23 - [trace] [flow:1f453d6c.db47e3] 765a34d6.999204 | debug |
14 Oct 23:47:23 - [trace] [flow:1f453d6c.db47e3] 240efecd.dcff2a | function |
14 Oct 23:47:23 - [trace] [flow:1f453d6c.db47e3] 25b3a4d3.5b7de4 | mongodb in |
14 Oct 23:47:23 - [trace] [flow:1f453d6c.db47e3] b466d287.dd4bd8 | function |
14 Oct 23:47:23 - [trace] [flow:1f453d6c.db47e3] 4bb24e3f.754308 | function |
14 Oct 23:47:23 - [trace] [flow:1f453d6c.db47e3] c54d19dc.2a4f18 | mongodb in |
14 Oct 23:47:23 - [trace] [flow:1f453d6c.db47e3] b87eacee.463238 | inject |
14 Oct 23:47:23 - [trace] [flow:1f453d6c.db47e3] 7557bd83.25f52c | switch |
14 Oct 23:47:23 - [trace] [flow:1f453d6c.db47e3] 4d5a6ba5.c87f94 | debug |
14 Oct 23:47:23 - [trace] [flow:1f453d6c.db47e3] 57f7b395.1be8b4 | function |
14 Oct 23:47:23 - [trace] [flow:1f453d6c.db47e3] 249ebb07.ecd3bc | function |
14 Oct 23:47:23 - [trace] [flow:1f453d6c.db47e3] 438e5c0b.9080ec | debug |
14 Oct 23:47:23 - [trace] [flow:1f453d6c.db47e3] 78abbbe2.3ae6bc | uuid |
14 Oct 23:47:23 - [trace] [flow:1f453d6c.db47e3] 695404fb.1371ec | function |
14 Oct 23:47:23 - [trace] [flow:1f453d6c.db47e3] eb84fbd7.725308 | mongodb out |
14 Oct 23:47:23 - [trace] [flow:1f453d6c.db47e3] 394d58b0.a247a | mongodb out |
14 Oct 23:47:23 - [trace] [flow:1f453d6c.db47e3] 42cfca84.96cfe4 | switch |
14 Oct 23:47:23 - [trace] [flow:1f453d6c.db47e3] a4555540.af55d8 | function |
14 Oct 23:47:23 - [trace] [flow:1f453d6c.db47e3] f5661a5e.99b128 | mongodb in |
14 Oct 23:47:23 - [trace] [flow:1f453d6c.db47e3] d6a01b33.9c70c8 | function |
14 Oct 23:47:23 - [trace] [flow:1f453d6c.db47e3] a0acb716.a05a4 | inject |
14 Oct 23:47:23 - [trace] [flow:1f453d6c.db47e3] b54c1e9d.c64e18 | function |
14 Oct 23:47:23 - [trace] [flow:1f453d6c.db47e3] aee6c107.3701c | mongodb in |
14 Oct 23:47:23 - [trace] [flow:1f453d6c.db47e3] 36b8a350.851acc | debug |
14 Oct 23:47:23 - [trace] [flow:1f453d6c.db47e3] 69bd9e3f.4a24e8 | debug |
14 Oct 23:47:23 - [trace] [flow:1f453d6c.db47e3] ------------------|--------------|-----------------
14 Oct 23:47:23 - [trace] runtime event: {"id":"runtime-state","retain":true}
14 Oct 23:47:23 - [info] Started flows
14 Oct 23:47:25 - [trace] comms.open BkNAWN07z6mZaPBGZoPw05BMCZOsCFZA
```

This is the section relating to the flow starting, and the rest of the log.

**Comment by Ryonez** [comment]
@knolleary
What info is needed? This is still happening and I provided the trace above.

**Comment by knolleary** [comment]
I was assuming you'd use the trace yourself to look for anything odd around what nodes are being stopped and started - the sample of trace you've provided doesn't include all the 'starting' and 'stopping' messages.

Looking at the flow you've shared, I can see a simple path where the msg from the HTTP In node is passed straight to the HTTP Response node without setting a payload - which would cause the behaviour you describe. This happens when `msg.req.query.request` is equal to `update`.

Where in your flow have you put Debug nodes to try to understand what is happening?

For example, I would suggest putting one on the output of every node you have wired to the HTTP Response node. You can then use that to see which one is passing a message to the response node in the case where you don't get the response you expect.

**Comment by Ryonez** [comment]
> the sample of trace you've provided doesn't include all the 'starting' and 'stopping' messages.

That's correct, I identified the startup of the flow and posted that. As the issue is with it starting, i didn't look for stopping sections.

> Looking at the flow you've shared, I can see a simple path where the msg from the HTTP In node is passed straight to the HTTP Response node without setting a payload - which would cause the behaviour you describe. This happens when msg.req.query.request is equal to update.

I don't see how that would cause the error.

The setup is as such. My phone makes a "RetrieveFullList" request regularly. That the only request it makes, until that query returns something. If something is return, the phone processes it and then (and only then) preforms an update request to red with what's needed to red to update whatever. the switch node links "update" back to the http response node only to close the com line. This doesn't cause an error, and if I don't have it, I get a timeout error on the phone.

When red restarts, the http response node is broken. it doesn't send json data to the phone at all, and nothing works. It's not until I do something like toggle a debug node that it'll all start working again.

> For example, I would suggest putting one on the output of every node you have wired to the HTTP Response node.

I stuck a node were things should be processing the request. According to the debug, *nothing* is even seen on the http in node. Phone shows an error for invalid json data. (red is just parroting the "RetrieveFullList" request back to the phone.)

**Comment by knolleary** [comment]
Closing as part of a tidy-up of the issue tracker — this has been inactive for a long time and much has changed in Node-RED since. If it's still relevant on a current version, please open a fresh issue with up-to-date details and we'll take a look. Thanks for raising it.
